### PR TITLE
Support complex number list in JIT

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51145 Support complex number list in JIT**

Differential Revision: [D26154025](https://our.internmc.facebook.com/intern/diff/D26154025)

### CODE REVIEW

File: aten/src/ATen/core/type.cpp [hunk 1]
```diff
@@ -1120,12 +1124,12 @@ std::string ClassType::getForwardPreHookErrorMessage(int pre_hook_idx) const {
   const FunctionSchema& forward_schema = getMethod("forward").getSchema();
   std::string input_types = getSchemaInputTypesString(forward_schema);
   const std::vector<Argument>& forward_args = forward_schema.arguments();
-
```
Comment (SplitInfinity) [inline aten/src/ATen/core/type.cpp:1123]: There seem to be a lot of formatting changes below this point. Can you undo them?

**Review by SplitInfinity — COMMENTED** [review]
I'm not sure if this is ready for review (I didn't check if you requested review already, oops) but there are a few distracting formatting changes, and there should be Python tests that use lists of complex numbers.

File: aten/src/ATen/core/type.cpp (hunk 1 above)
Comment (anjali411) [reply aten/src/ATen/core/type.cpp:1123]: they seem inevitable

**Comment by anjali411** [comment]
@SplitInfinity ready for review. Added the python test, but can't find a way to update that file without causing those formatting changes

File: aten/src/ATen/core/type.cpp (hunk 1 above)
Comment (anjali411) [reply aten/src/ATen/core/type.cpp]: done!

File: aten/src/ATen/test/ivalue_test.cpp [hunk 2]
```diff
@@ -49,6 +49,26 @@ TEST(IValueTest, Basic) {
...
+  auto foo12 = std::move(bar1);
+  ASSERT_EQ(foo1.use_count(), 3);
+  ASSERT_TRUE(foo12.isComplexDoubleList());
+  ASSERT_TRUE(bar1.isNone());
+  foo12 = IValue(c10::complex<double>(3, 4));
+  ASSERT_TRUE(foo12.isComplexDouble());
+  ASSERT_EQ(foo12.toComplexDouble(), c10::complex<double>(3,4));
+  ASSERT_EQ(foo1.use_count(), 2);
+  ASSERT_TRUE(baz1.toComplexDoubleVector() == std::vector<c10::complex<double>>({(3, 4), (3, -4), (5, 0)}));
+  IValue the_complex_list(
```
Comment (SplitInfinity) [inline aten/src/ATen/test/ivalue_test.cpp:68]: Seems a bit weird to call this a list when it is a tuple.

File: test/jit/test_complex.py [hunk 3]
```diff
@@ -14,6 +14,13 @@ def fn(a: complex):
 
         self.checkScript(fn, (3 + 5j,))
 
+    def test_complexlist(self):
+        def fn(a: List[complex], idx: int):
+            return a[idx]
+
+        input = [1j, 2, 3 + 4j, -5, -7j]
```
Comment (SplitInfinity) [inline test/jit/test_complex.py:21]: What about serialization of lists? Is that coming later?

**Review by SplitInfinity — COMMENTED** [review]
Nice!

File: torch/csrc/jit/ir/constants.cpp [hunk 4]
```diff
@@ -88,8 +88,8 @@ c10::optional<Value*> tryInsertConstant(
     n->i_(attr::value, val.toBool());
     n->output()->setType(BoolType::get());
   } else if (val.isList()) {
-    bool fast_path_list =
-        val.isBoolList() || val.isIntList() || val.isDoubleList();
+    bool fast_path_list = val.isBoolList() || val.isIntList() ||
+        val.isDoubleList() || val.isComplexDoubleList();
```
Comment (SplitInfinity) [inline torch/csrc/jit/ir/constants.cpp:92]: As I recall, complex literals are not supported in TorchScript yet, so this is not being tested.

File: aten/src/ATen/test/ivalue_test.cpp (hunk 2 above)
Comment (anjali411) [reply aten/src/ATen/test/ivalue_test.cpp:68]: true updated the name!

File: torch/csrc/jit/ir/constants.cpp (hunk 4 above)
Comment (anjali411) [reply torch/csrc/jit/ir/constants.cpp:92]: removed!

File: test/jit/test_complex.py (hunk 3 above)
Comment (anjali411) [reply test/jit/test_complex.py:21]: will add it in a follow-up PR

**Review by SplitInfinity — APPROVED** [review]
Nice

**Pull request closed without merging**